packages/instantsearch.js/src/lib/InstantSearch.ts

```diff
@@ -840,6 +842,100 @@ See documentation: ${createDocumentationLink({
   }
+
+  /**
+   * Update the parameters passed to the InstantSearch constructor.
+   */
+  public update(
+    options: Partial<InstantSearchOptions<TUiState, TRouteState>>
+  ): void {
+    if (this.indexName !== options.indexName) {
+      this.helper!.setIndex(options.indexName || '').search();
+    }
+
+    if (this.client !== options.searchClient) {
+      warning(
+        false,
+        // TODO: flavor-specific warning (search._flavor = 'vue') URL
+        'The `searchClient` parameter changed, which may cause more search requests than necessary. If this is an unwanted behavior, please provide a stable reference.'
+      );
+
+      if (!options.searchClient) {
+        throw new Error(withUsage('The `searchClient` option is required.'));
+      }
+
+      if (typeof options.searchClient.search !== 'function') {
+        throw new Error(
+          `The \`searchClient\` must implement a \`search\` method.
+  
+  See: https://www.algolia.com/doc/guides/building-search-ui/going-further/backend-search/in-depth/backend-instantsearch/js/`
+        );
+      }
+
+      // TODO: proper API for this too
+      // InstantSearch._algoliaAgents (or constructor)
+      // addAlgoliaAgents(props.searchClient, [
+      //   ...defaultUserAgents,
+      //   serverContext && serverUserAgent,
+      // ]);
+
+      this.mainHelper!.setClient(options.searchClient).search();
+    }
+
+    if (this.onStateChange !== options.onStateChange) {
+      this.onStateChange = options.onStateChange;
+    }
+
+    if (this._searchFunction !== options.searchFunction) {
+      this._searchFunction =
+        options.searchFunction ??
+        ((helper) => {
+          helper.search();
+        });
+    }
+
+    if (this._stalledSearchDelay !== options.stalledSearchDelay) {
+      this._stalledSearchDelay =
+        options.stalledSearchDelay ?? INSTANTSEARCH_STALLED_SEARCH_DEFAULTS;
+    }
+
+    // TODO: move dequal to a shared package?
+    // if (!dequal(this.future, props.future)) {
+    //   this.future = {
+    //     ...INSTANTSEARCH_FUTURE_DEFAULTS,
+    //     ...props.future,
+    //   };
+    // }
+
+    // TODO: implement update in middleware
+    // TODO: can this be simplified?
+    // if (!dequal(this._insights, options.insights)) {
+    //   this._insights = options.insights;
+    //   const existing = this.middleware.find(
+    //     (m) => m.instance.$$type === 'ais.insights' && m.instance.$$internal
+    //   );
+    //   if (options.insights && existing) {
+    //     // TODO: update existing middleware somehow (or maybe remount it)
+    //   } else if (options.insights && !existing) {
+    //     const insightsOptions =
+    //       typeof options.insights === 'boolean' ? {} : options.insights;
+    //     insightsOptions.$$internal = true;
+    //     this.use(createInsightsMiddleware(insightsOptions));
+    //   } else if (!options.insights && existing) {
+    //     this.unuse(existing.creator);
+    //   }
+    // }
```

Comment:
I'd love for middlewares to remove some of the load from this method. Maybe we could iterate over middlewares, call their `update()` method, and get a returned value (boolean? object?) which could tell `InstantSearch.update()` to unuse them.

Reply:
in this case probably `existing.update(options.insights)` and handling that somehow in the middleware indeed was best, wasn't sure whether the api was worth the coals